pkg/endpoint/policy.go

```diff
@@ -353,15 +363,71 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *LabelsMap,
 				e.cleanUnusedRedirects(owner, e.L4Policy.Ingress, c.L4Policy.Ingress)
 				e.cleanUnusedRedirects(owner, e.L4Policy.Egress, c.L4Policy.Egress)
 			}
-			l4Rm, l4Add, err = e.applyL4PolicyLocked(owner, labelsMap, e.L4Policy, c.L4Policy)
+			rulesAdd, l4Rm, err = e.applyL4PolicyLocked(owner, labelsMap, e.L4Policy, c.L4Policy)
 			if err != nil {
 				// This should not happen, and we can't fail at this stage anyway.
-				e.getLogger().Fatal("L4 Policy application failed")
+				e.getLogger().WithError(err).Fatal("L4 Policy application failed")
 			}
 		}
-		e.L4Policy = c.L4Policy // Reuse the common policy
+		// Reuse the common policy, will be used in lxc_config.h (CFG_L4_INGRESS and CFG_L4_EGRESS)
+		e.L4Policy = c.L4Policy
 		e.LabelsMap = labelsMap // Remember the set of labels used
+
+		// We need to know which ones are L4-only by checking
+		// if there's any L4-only rules that doesn't match a L3-L4 rule.
+		if c != nil && c.L4Policy != nil && c.L4Policy.Ingress != nil {
+			for _, l4Filter := range c.L4Policy.Ingress {
+				found := false
+				l4RuleCtx, l7RuleCtx := policy.ParseL4Filter(l4Filter)
+				for _, l4RuleContexts := range rulesAdd {
+					if _, found = l4RuleContexts[l4RuleCtx]; found {
+						break
+					}
+				}
+				if !found {
+					if _, ok := rulesAdd[policy.InvalidIdentity]; !ok {
+						rulesAdd[policy.InvalidIdentity] = policy.NewL4RuleContexts()
+					}
+					l7RuleCtx.L4Installed = true
+					rulesAdd[policy.InvalidIdentity][l4RuleCtx] = l7RuleCtx
+				}
+			}
+		}
+
 		changed = true
+
+		// Only remove the CT entries of the rules that were successfully
+		// removed (added to l4Rm) and not successfully re-added (added to rulesAdd)
+		for secID, ruleCtxsRm := range l4Rm {
+			if addL4RuleCtxs, ok := rulesAdd[secID]; ok {
+				for addl4RuleCtx, addl7RuleCtx := range addL4RuleCtxs {
+					rm, ok := ruleCtxsRm[addl4RuleCtx]
+					if rm.L4Installed && ok && !(addl7RuleCtx.L4Installed) {
+						if _, ok := rulesRm[secID]; !ok {
+							rulesRm[secID] = policy.NewL4RuleContexts()
+						}
+						rulesRm[secID][addl4RuleCtx] = addl7RuleCtx
+					}
+				}
+			} else {
+				// If the L3 rule was not re-added then we clean up all CT
+				// entries only based on L3.
+				if _, ok := rulesRm[secID]; !ok {
+					rulesRm[secID] = policy.NewL4RuleContexts()
+				}
+				for rmL4RuleCtx, rmL7RuleCtx := range ruleCtxsRm {
+					removed, ok := rulesRm[secID][rmL4RuleCtx]
+					rmL7RuleCtx.L4Installed = rmL7RuleCtx.L4Installed || (removed.L4Installed && ok)
+					rulesRm[secID][rmL4RuleCtx] = rmL7RuleCtx
+				}
+			}
+		}
+	} else {
```

Comment:
Could this case go first? By the time I reached the else, I forgot what the branch was on :/

Reply:
Done